[Debt] Migrate to vite

esizer opened the pull request:
:robot: Resolves #9840 

## 👋 Introduction

Migrates from webpack to vite.

## 🕵️ Details

### Watch vs Dev/Build

Thanks to @substrae we do not have to sacrifice the watch command. However it now works a little different. When running `pnpm run dev/build` everything works as you should expect. But in order to watch, we use vite's dev server for HMR. This means, when running watch, you should be using `http://localhost:3000` (note the port number).

Now, when developing, you no longer need to refresh the browser! Vite will reload the browser and reload the components. It does this on demand so you may notice some flashing and temporary error pages but I think this is a good trade off.

### Barrel exports

It may be time to revisit removing all barrel exports again since vite [discourages it in their performance section](https://vitejs.dev/guide/performance#avoid-barrel-files). We do not do this too much but I think we still have some.

## :broken_heart: Breaking Changes

### HTML Template

This has been moved to the root (expected by vite).

### Environment variables

This has changed significantly. We no longer use the `config.ejs` file for runtime variables since vite supports `ejs` in the `html` template. 

To add an environment variable, you still add it to the `.env` but instead of updated the `config.ejs` for runtime vars, you need to add it to:

- `vite.config.ts`
- `vite-env.d.ts`
- `index.html` (for runtime only)

To access an environment variable, just reference it like a global constant. We no longer need to use the `process.env` anymore.

## :star: Bonus

### Storybook

This seems to also resolve #10557 so it would benefit us for you to test that as well :smile: 

### Codegen watch

I was also able to figure out why codegen watch was broken and fixed that here as well since it was only 2 character change :sweat_smile: 

## 🧪 Testing

1. Copy over new `.env.example` items (`DEV_APP_URL`, `DEV_OAUTH_POST_LOGIN_REDIRECT`, `API_HOST`) 
2. Clean out your modules `make clean-modules`
3. Reinstall deps `pnpm i`
4. Build `pnpm run build`
5. Check to ensure app still works

vd1992 commented:
😮 

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/731353e1-f43f-4c9e-bcb5-3356d94648d9)

esizer commented:
> 😮

:zap:

esizer commented on apps/web/src/pages/AccessibilityStatementPage/AccessibilityStatementPage.tsx:
```diff
@@ -274,7 +273,7 @@ export const Component = () => {
         crumbs={crumbs}
       />
-      <div data-h2-container="base(center, large, x1) p-tablet(center, large, x2)">
+      <div data-h2-container="base(center, large, x1) p-tablet(center, large, x14)">
```
Note to self: I was doing some testing with HMR and h2... need to fix this.

esizer commented on apps/web/index.html:
```diff
@@ -2,11 +2,8 @@
 <html lang="en" data-h2>
   <head>
     <meta charset="UTF-8" />
-    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
```
I think this was a mistake. We probably want to bring this back :thinking:

substrae commented:
@esizer So I was fiddling around tonight and I can't seem to get `pnpm run watch` to compile successfully at all. `build` seems to work as expected. Running `watch` hangs here:

```
@gc-digital-talent/web:watch:   at async loadConfigFromFile (/var/www/html/node_modules/vite/dist/node/chunks/dep-BKbDVx1T.js:69764:28)
@gc-digital-talent/web:watch: 
@gc-digital-talent/web:watch: 
✔ Parse Configuration
@gc-digital-talent/graphql:watch: ✔ Generate outputs
@gc-digital-talent/graphql:watch: 
```

That said, I was able to get one or two successful runs using Hydrogen by applying it to the Vite plugin array this way:

```
function runHydrogen(): Plugin {
  return {
    name: "Hydrogen Vite Plugin",
    configResolved() {
      hydrogen_build();
    },
  };
}
```

Typescript wasn't happy about the way I was importing the function from `node_modules`, but that's probably just my own ignorance. It compiled as expected and didn't seem to create the infinite loop, however I noticed I also couldn't trigger watch to rebuild either, so perhaps there are errors not being output in the console?

I'm trying all of this from the `Maintenance` container, so I'm going to try installing `pnpm` directly into my WSL instance and running it that way to see if there are differences.

substrae commented:
I opted to try and troubleshoot using a vanilla React+Vite project, and running Hydrogen on `configResolved` and `handleHotUpdate` seems to accomplish what we're looking for, with the caveat that it runs twice on hot reload, something I can't seem to resolve with my limited knowledge.

esizer commented:
> I opted to try and troubleshoot using a vanilla React+Vite project, and running Hydrogen on `configResolved` and `handleHotUpdate` seems to accomplish what we're looking for, with the caveat that it runs twice on hot reload, something I can't seem to resolve with my limited knowledge.

I'm guessing it is acting funky because [configResolved](https://vitejs.dev/guide/api-plugin.html#configresolved) is meant to allow you to resolve certain config values and shouldn't be use to run build scripts.

I tried this config and cannot actually get hydrogren to ever build :-1: It reloads the page and the infinite re-build is fixed but that is likely because hydrogen is not writing to the file now.

substrae commented:
@esizer Given that it runs okay on a vanilla install, it not running at all is likely unique to our Vite config or project setup. 

I can't really troubleshoot until I can get our app running locally again, but the WSL/Docker permission conflicts are preventing me from running commands. Vite's watch command doesn't seem to want to complete inside of Docker, so I'm forced to run it outside of Docker by installing `pnpm` on my WSL distro, but then that command complains about Docker owning the `node_modules`. Happy to try and pair on resolving that, but it's unrelated to this PR specifically.

petertgiles commented on apps/web/vite.config.ts:
Wow wow wow!
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/c89dad8e-6dec-422c-9018-5201892b6213)
Fresh build, no maintenance container.

petertgiles commented on api/config/cors.php:
```diff
@@ -15,7 +15,7 @@
     */
 
-    'paths' => ['api/*', 'sanctum/csrf-cookie'],
+    'paths' => ['api/*', 'sanctum/csrf-cookie', 'graphql'],
```
Also we'll need to add admin/graphql.  Any others?  :thinking:

esizer replied:
Thanks for catching that. Added in [add protected endpoint to cors](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10611/commits/63ad319cbf7d166e70d025f497dea3b3f660eb38)

esizer commented:
> @esizer Given that it runs okay on a vanilla install, it not running at all is likely unique to our Vite config or project setup.
> 
> I can't really troubleshoot until I can get our app running locally again, but the WSL/Docker permission conflicts are preventing me from running commands. Vite's `watch` command doesn't seem to want to complete inside of Docker, so I'm forced to run it outside of Docker by installing `pnpm` on my WSL distro, but then that command complains about Docker owning the `node_modules`. Happy to try and pair on resolving that, but it's unrelated to this PR specifically.

> @esizer Given that it runs okay on a vanilla install, it not running at all is likely unique to our Vite config or project setup.
> 
> I can't really troubleshoot until I can get our app running locally again, but the WSL/Docker permission conflicts are preventing me from running commands. Vite's `watch` command doesn't seem to want to complete inside of Docker, so I'm forced to run it outside of Docker by installing `pnpm` on my WSL distro, but then that command complains about Docker owning the `node_modules`. Happy to try and pair on resolving that, but it's unrelated to this PR specifically.

Did you just do `pnpm create vite my-react-app --template react-ts`? And where were you building the css file to? I can get it working by swapping the order in your example but it does not fix the infinite re-build.

petertgiles commented on apps/web/src/components/PoolFilterInput/PoolFilterInput.tsx:
```diff
@@ -3,7 +3,10 @@ import debounce from "lodash/debounce";
 import { Combobox } from "@gc-digital-talent/forms";
-import { PoolFilterInput, Scalars } from "@gc-digital-talent/graphql";
+import type {
+  PoolFilterInput as TPoolFilterInput,
```
Why the `T` prefix.  That's unconventional, no?

esizer replied:
Unconventional for us, but not typescript in general. 

However, this is only required to avoid name collision. I couldn't think of a better name for the component so I just aliased it with a prefix. I would love to avoid this so I am more than happy to take suggestions! :smiley:

petertgiles replied:
Oh yeah, I see it now.  A better name doesn't come to mind so this is as good as any.

esizer replied:
Yeah, I'm kind of disappointed this did not get caught when building with webpack :confounded:

substrae commented:
@esizer 
* I ran `npm create vite@latest` and chose `react`, 
* Installed Hydrogen using `npm install --save @hydrogen-css/hydrogen`
* Ran `npx h2-init` and set the `output` to `src/assets`
* Imported Hydrogen into the Vite config using `import {hydrogen_build} from '@hydrogen-css/hydrogen/lib/build'`
* Added the plugin as above, then added the function to the plugin array after `react`

petertgiles commented on documentation/environment-variables.md:
```diff
@@ -12,11 +12,11 @@ The frontend [React](https://reactjs.org/) projects uses environment variables i
 ### Run-Time Variables
 
-For deployment in production, there needs to be a way to change variables in the program without rebuilding. This allows us to use environment variables as "feature flags", which enable/disable specific features with the same deployed code artifact. This is done using [envsubst](https://www.gnu.org/software/gettext/manual/html_node/envsubst-Invocation.html) to fill environment variables into the  ...
+For deployment in production, there needs to be a way to change variables in the program without rebuilding. This allows us to use environment variables as "feature flags", which enable/disable specific features with the same deployed code artifact. This is done using [envsubst](https://www.gnu.org/software/gettext/manual/html_node/envsubst-Invocation.html) to fill environment variables into the  ...
```
config.js is gone now, right?

petertgiles replied:
I guess this subsequent change is TBD once the deployment process is updated.

esizer replied:
Yup! Will that cause any issues I am unaware of?

petertgiles replied:
I don't think so.  We should be able to call envsubst on the index.html file the same way we do currently on the config.js file.

One enhancement I've been thinking about is that we can get rid of all the "filterUnusable" nonsense if we can emit two index.html files - one with the process.env substitutions completed for local dev and one with the variables still there for deployment.  That might be more than you want to deal with right now, though.

esizer replied:
Yeah, that sounds possible. I think we could do something with that using the [template option](https://github.com/vbenjs/vite-plugin-html#:~:text=entry%20file%20path-,template,-string) in the html plugin along with the [modes](https://vitejs.dev/guide/env-and-mode.html#modes) from vite :thinking: 

```ts
createHtmlPlugin({
  template: modes === "production" ? "prod.index.html" : "dev.index.html"
  // ...
})
```

petertgiles replied:
That's cool.  I don't want to blow out the scope though.  I'll let you decided whether it's worth trying for or consider it later.

petertgiles replied:
Without a config file to copy, you'll need to update the post-deploy script.  I know you don't have access to Azure but the web container should be a pretty good facsimile.
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/5f53c5ae-7963-4d55-8c8a-5f2cb4d174c6)

esizer replied:
I think I got it in: [remove references to config.js](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10611/commits/cc73ca5bbb105d7164b8ef5d68a4c71402e9b052) :sweat_smile:

esizer replied:
Alright, I thanks for walking me through that. I think it should be good after [fix post deploy env substitution](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10611/commits/b74f6b515a37b9569ad3db68874cf4361a7b4e21)

petertgiles commented on packages/client/src/exchanges/protectedEndpointExchange.ts:
```diff
@@ -27,7 +28,7 @@ const protectedEndpointExchange: Exchange =
           // we're updating the operation on the fly
           // eslint-disable-next-line no-param-reassign
-          op.context.url = protectedUrl;
+          op.context.url = protectedUrl ?? "";
```
This null coalesce is unnessary because of the ternary above, right?

esizer replied:
Good catch, also thanks for the help on the dev server stuff. This was fixed as part of that fix.

[Fix dev server api](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10611/commits/3e10072ff0d2c83756686d7406143ce6b4c50a56)

petertgiles commented on apps/web/package.json:
```diff
@@ -2,12 +2,13 @@
+    "build": "vite build --mode production --minify",
     "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist",
-    "dev": "webpack --config webpack.dev.js",
+    "dev": "vite build --sourcemap --mode=development",
```
Maybe I never noticed this before because I always used watch mode but codgen isn't caching when I run this.

esizer replied:
Oh yeah, we disabled it because it seemed to be over-caching :cry: Maybe worth another pass to see if we can figure that out. 

https://github.com/GCTC-NTGC/gc-digital-talent/blob/acd32b7e1f97a0e0138a82374e8d8fbfd9caa176/turbo.json#L27

petertgiles commented on apps/web/index.html:
```diff
@@ -68,18 +65,64 @@
+      const filterUnusable = (value) =>
+        !value.startsWith("$") && value.length > 0 ? value : undefined;
+
+      const data = new Map([
```
I love that you were able to roll this in and remove that extra network roundtrip!

petertgiles reviewed (commented):
This is looking very cool.  I'll mourn the (hopefully temporary) loss of watch mode but overall this seems like a big step forward.  Let me know if you want help with the Azure deployment part of this.

substrae commented:
Created a discussion over on Vite's repo to see if we can crowdsource some help from folks who have more experience: https://github.com/vitejs/vite/discussions/17470

vd1992 commented:
Just trying out install and building, things seem to work!

`pnpm run dev:fresh`

About ~70% reduction in time

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/800c894c-78a7-43dc-99a7-8816145dea63)

vd1992 commented:
Some of the tests are failing due to missing localization

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/12fe7f70-47e7-4cbb-adc7-2a799e541f75)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/a1f930b8-cdb2-482d-bd99-a692b5f8915c)

vd1992 commented:
Being strange in quite a few places

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/a7bfb3da-0910-4a82-a28d-6baf38aa80a1)

esizer commented:
> Some of the tests are failing due to missing localization

Yeah, it is so odd. I have been trying to find out why some are missing but can't figure it out! :confounded:

esizer commented on packages/client/env.d.ts:
We could just pass these in as props so we don't need to have the declaration here :thinking:

vd1992 commented:
Refreshing on various pages outputs this in the console, doesn't break the page as far as I can tell though

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/a56f3cdf-f3e5-43d1-a227-e781e7c2dd13)

vd1992 commented:
Admin side menu rendering

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/2a116a69-a942-4fbe-a258-fd93c9119a77)

vd1992 commented:
The spinner appears to no longer rotate. This thing

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/98f065fa-13b0-4549-942e-740777f62f95)

And in some cases doesn't appear at all?

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/10ba890c-f95b-40e1-85c0-6bdfa71a7e3c)

vd1992 commented:
Dev server will need some bespoke environment values like the login redirects, otherwise you cannot login 

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/7760d960-6623-4f3c-98a8-e466dd18f03d)

esizer commented:
> Refreshing on various pages outputs this in the console, doesn't break the page as far as I can tell though

Looks like there was an issue with `framer-motion` loading async. I updated it to be synchronous but split it off to a chunk to still keep the main bundle size savings. All animations should be working now.

[remove async animation loading, split off to chunk](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10611/commits/5be2315a9766bb903cefa807665e359f6ff76b83)

> Dev server will need some bespoke environment values like the login redirects, otherwise you cannot login

I'm a little confused. The variables do exist and your screenshot seems to be showing the expected URL :thinking:

vd1992 commented:
Navigating in the dev server isn't working well for me, I get various things output to the console and page crashes. Page works after refreshing though. 
Guess it isn't loading things properly so it crashes the first time only. 

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/87ead93e-b4a6-4cfb-81a6-7fc0eddf7a95)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/dd4f8b9e-8073-457f-a45b-2fa04276c80c)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/916b7920-78f1-4d68-be98-69f0e5f281bd)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/b4d5af31-9601-40d8-a626-46dc0ab14b9f)

vd1992 commented:
Something similar when running storybook, it has loading errors but resolves within ~10 seconds and seems to work fine otherwise

vd1992 commented:
> > Navigating in the dev server isn't working well for me, I get various things output to the console and page crashes. Page works after refreshing though.
> > Guess it isn't loading things properly so it crashes the first time only.
> 
> This is [entirely expected](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10611#:~:text=It%20does%20this%20on%20demand%20so%20you%20may%20notice%20some%20flashing%20and%20temporary%20error%20pages%20but%20I%20think%20this%20is%20a%20good%20trade%20off.), vite will transform and load stuff on demand so you will see errors pop in and out but they should resolve themselves fairly quickly.

Okay, hopefully it works more smoothly for others

vd1992 commented:
Okay, I think things look better now, let us get some other eyes on this 👀

mnigh reviewed (approved):
🥳 this is so great!

❓ perhaps it is worth enabling the Chromatic workflow and running it on this PR as well, if the intent is for this PR to resolve that issue?

> [!IMPORTANT]  
> my approval is limited to local testing (similar to @vd1992's implicit approval). i would wait to merge until @petertgiles is confident with deploying and testing on remote servers.

petertgiles commented:
Build error in router.tsx :cry:

esizer commented:
> Build error in router.tsx 😢

Not really sure why this occurred :confused: . But it is now fixed with [split experience form pages into own files](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10611/commits/1184c9f48fa37cc68d6ae55b8f70a01d03f38918)

vd1992 commented:
Checking things out again to see what changed

What exactly is the difference between `run dev` and `run build` ? Can't quite tell the difference despite it correctly outputting "building for production/development"
Takes about the same time to run and the request total (request transferred size in browser load) stays the same too. Or am I missing the details?

`run watch` has a much higher request total on the other-hand, when the page loads 😛 
With the exceptions of a few pages it is working a bit better

On the subject of request total, significantly higher number of requests isn't it? Might need this sooner than later https://github.com/GCTC-NTGC/gc-digital-talent/issues/8350

esizer commented:
> What exactly is the difference between `run dev` and `run build` ? Can't quite tell the difference despite it correctly outputting "building for production/development"

It is subtle but important, dev command changed the `NODE_ENV` which will allow us to modify env vars if we want. In addition to that, it will create source maps for better debugging. These source maps are external so you shouldn't really notice them in the actual bundle size.

Dev mode strips all this out to create a more optimized build for production.

As for watch, the request number is so high becuase of how the HMR works. It dynamically loads chunks of code into the browser on demand and does this often (on every file save) to update the page.

vd1992 commented:
Thanks for answering

I noticed that there is a `pnpm run build-storybook` that outputs to `/gc-digital-talent/apps/web/storybook-static`
How do I use it or reach it lol

Playing around with the various scripts, think we had a bug in one of them
The web one should point at `/web` right?

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/878100af-e4bb-4931-b1f0-6c079d1d2a0e)

`pnpm run lint` keeps crashing for me, anyone else having issues with it? Same with running jest. Process killed or WSL crash

Playwright still works well, as does codegen and tsc

esizer commented:
> I noticed that there is a `pnpm run build-storybook` that outputs to `/gc-digital-talent/apps/web/storybook-static`
> How do I use it or reach it lol

It is set in the chromatic workflow. The folder name is the standard name so it just gets picked up there.

https://github.com/GCTC-NTGC/gc-digital-talent/blob/5844ad8deac2e58da7faa335e007aad16627286a/.github/workflows/chromatic.yml#L67

> Playing around with the various scripts, think we had a bug in one of them
> The web one should point at `/web` right?

It is a little confusing but, it is pointing to the correct location. The script is chekcing web but it is run in that package since it uses the `cli.ts` file in that package.

https://github.com/GCTC-NTGC/gc-digital-talent/blob/5844ad8deac2e58da7faa335e007aad16627286a/packages/i18n/package.json#L16

petertgiles reviewed (changes requested):
This is amazing.  😍   Locally, lint, Playwright, watch, storybook all work great.

I'm tracking only two "only in Azure" issues here.

esizer commented on .github/workflows/chromatic.yml:
Oh! I actually ran into this the other day on a different branch. WE need to run `php artisan cache:clear` to fix this I think.

If that does not work, then maybe there is something else I missed :sweat_smile:

petertgiles replied:
I went nuclear and reset the whole database in dev and it fixed it.  😅

petertgiles replied:
Possibly related as well:

petertgiles commented on .github/workflows/chromatic.yml:
This is still happening after the reset.  Happens in both Edge and Chrome.  😕

petertgiles replied:
Issue was that there were no communities seeded.  I'll make a bug but out of scope for this PR.

petertgiles replied:
https://github.com/GCTC-NTGC/gc-digital-talent/issues/10767
